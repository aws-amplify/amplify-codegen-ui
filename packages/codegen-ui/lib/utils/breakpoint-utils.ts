/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { StudioComponent } from '../types';

export const breakpointSizes = ['base', 'small', 'medium', 'large', 'xl', 'xxl'] as const;
export type BreakpointSizeType = typeof breakpointSizes[number];
export const bpWeights: Record<BreakpointSizeType, number> = {
  base: 0,
  small: 1,
  medium: 2,
  large: 3,
  xl: 4,
  xxl: 5,
};

/**
 * sorts the breakpoints to the following order
 *
 * ['base', 'small', 'medium', 'large', 'xl', 'xxl']
 */
export const sortBreakpoints = (bs: BreakpointSizeType[]): BreakpointSizeType[] => {
  return bs.sort((lhs, rhs) => {
    return bpWeights[lhs] - bpWeights[rhs];
  });
};

export const getBreakpoints = (component: StudioComponent & Required<Pick<StudioComponent, 'variants'>>) => {
  const breakpoints = component.variants.reduce<BreakpointSizeType[]>((acc, variant) => {
    if (variant.variantValues?.breakpoint) {
      acc.push(variant.variantValues.breakpoint as BreakpointSizeType);
    }
    return acc;
  }, []);
  return sortBreakpoints(breakpoints);
};
