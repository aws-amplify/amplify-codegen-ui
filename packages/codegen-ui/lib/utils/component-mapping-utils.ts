/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent, StudioComponentChild } from '../types';

/**
 * Helper to recurse through the component tree and build the name to type mapping.
 */
export const buildComponentNameToTypeMap = (
  component: StudioComponent | StudioComponentChild,
): Record<string, string> => {
  const localMap: Record<string, string> = {};
  if (component.name) {
    localMap[component.name] = component.componentType;
  }
  if (component.children) {
    Object.entries(
      component.children
        .map((child) => buildComponentNameToTypeMap(child))
        .reduce((previous, next) => {
          return { ...previous, ...next };
        }, {}),
    ).forEach(([name, componentType]) => {
      localMap[name] = componentType;
    });
  }
  return localMap;
};
