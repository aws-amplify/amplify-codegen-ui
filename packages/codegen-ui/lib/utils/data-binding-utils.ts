/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import {
  ActionStudioComponentEvent,
  StudioComponent,
  StudioComponentChild,
  StudioComponentEvent,
  StudioForm,
} from '../types';

const dataActions = [
  'Amplify.DataStoreCreateItemAction',
  'Amplify.DataStoreUpdateItemAction',
  'Amplify.DataStoreDeleteItemAction',
];

export function formRequiresDataApi(form: StudioForm): boolean {
  return form.dataType.dataSourceType === 'DataStore';
}

export function componentRequiresDataApi(component: StudioComponent): boolean {
  // data-bound collections have their data dependency persisted as a collection property
  if (component.componentType === 'Collection' && component.collectionProperties) {
    return Object.values(component.collectionProperties).length > 0;
  }

  // data-bound components need to be searched recursively for datastore actions
  return hasDataAction(component);
}

function isActionEvent(event: StudioComponentEvent): event is ActionStudioComponentEvent {
  return 'action' in event && !!event.action;
}

function hasDataAction(component: StudioComponentChild): boolean {
  const actions = Object.values(component.events || {})
    .filter(isActionEvent)
    .map((e) => e.action);
  if (actions.length && actions.some((a) => dataActions.includes(a))) {
    return true;
  }

  return (component.children || []).some(hasDataAction);
}
