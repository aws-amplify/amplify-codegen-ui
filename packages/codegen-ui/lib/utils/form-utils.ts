/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { StudioForm } from '../types';

export const shouldIncludeCancel = ({ cta }: StudioForm): boolean => {
  // first check if excluded is added in explicitly
  if (cta?.cancel && 'excluded' in cta.cancel && cta.cancel.excluded) {
    return false;
  }
  // if we find that cta has cancel defined (ex. an empty object)
  // we include the cancel button
  if (cta && cta.cancel) {
    return true;
  }

  // otherwise we return false
  return false;
};
