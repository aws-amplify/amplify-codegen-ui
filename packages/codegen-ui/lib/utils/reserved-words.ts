/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

/**
 * uesd so a reserved word is not generated in the component
 * src: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#reserved_words
 */
export const reservedWords = [
  'await',
  'break',
  'case',
  'catch',
  'class',
  'const',
  'continue',
  'debugger',
  'default',
  'delete',
  'do',
  'else',
  'enum',
  'export',
  'extends',
  'false',
  'finally',
  'for',
  'function',
  'if',
  'implements',
  'import',
  'in',
  'instanceof',
  'interface',
  'let',
  'new',
  'null',
  'package',
  'private',
  'protected',
  'public',
  'return',
  'super',
  'switch',
  'static',
  'this',
  'throw',
  'try',
  'True',
  'typeof',
  'var',
  'void',
  'while',
  'with',
  'yield',
  'React',
  // are variables names used wtihin the form
  'DataStore',
  'initialValues',
  'initialData',
  'onSubmit',
  'onCancel',
  'onChange',
  'onValidate',
  'fetchByPath',
  'validateField',
  'overrides',
];
