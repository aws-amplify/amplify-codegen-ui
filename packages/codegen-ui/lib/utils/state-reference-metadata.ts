/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent, StudioComponentProperty, StateReference } from '../types';
import { getComponentFromComponentTree } from './component-tree';

/**
 * TODO: UNIT TESTS ON ME
 */

export type StateReferenceMetadata = {
  reference: StateReference;
  dataDependencies: string[];
};

/**
 * Component Structure Functions
 */

export function computeStateReferenceMetadata(
  component: StudioComponent,
  stateReferences: StateReference[],
): StateReferenceMetadata[] {
  return stateReferences.map((reference) => {
    const dataDependencies = computeDataDependenciesForStateReference(component, reference);
    return { reference, dataDependencies };
  });
}

function reduceDataDependencies(dataDependencies: string[]): string[] {
  return [...new Set(dataDependencies)];
}

/**
 * Property Functions
 */

function computeDataDependenciesForStudioComponentProperty(property: StudioComponentProperty): string[] {
  return reduceDataDependencies(
    ([] as string[])
      .concat(
        'concat' in property && property.concat
          ? property.concat.flatMap(computeDataDependenciesForStudioComponentProperty)
          : [],
      )
      .concat('userAttribute' in property && property.userAttribute ? ['authAttributes'] : [])
      .concat(
        'bindingProperties' in property && property.bindingProperties ? [property.bindingProperties.property] : [],
      )
      .concat('collectionBindingProperties' in property && property.collectionBindingProperties ? ['items'] : [])
      .concat(
        'condition' in property && property.condition && 'property' in property.condition && property.condition.property
          ? [property.condition.property]
          : [],
      )
      .concat(
        'condition' in property && property.condition && 'then' in property.condition && property.condition.then
          ? computeDataDependenciesForStudioComponentProperty(property.condition.then)
          : [],
      )
      .concat(
        'condition' in property && property.condition && 'else' in property.condition && property.condition.else
          ? computeDataDependenciesForStudioComponentProperty(property.condition.else)
          : [],
      ),
  );
}

/**
 * State Functions
 */

function computeDataDependenciesForStateReference(
  component: StudioComponent,
  stateReference: StateReference,
): string[] {
  const { componentName, property } = stateReference;
  const namedComponent = getComponentFromComponentTree(component, componentName);
  const propertyDefinition = namedComponent.properties[property];
  if (propertyDefinition) {
    return computeDataDependenciesForStudioComponentProperty(propertyDefinition);
  }
  return [];
}
