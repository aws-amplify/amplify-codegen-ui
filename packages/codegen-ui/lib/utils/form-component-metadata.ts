/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { camelCase } from 'change-case';

import {
  FormDefinition,
  FormMetadata,
  FieldConfigMetadata,
  StudioForm,
  FieldValidationConfiguration,
  FormDefinitionElement,
  FormDefinitionInputElement,
} from '../types';
import { reservedWords } from './reserved-words';

export const getFormFieldStateName = (formName: string) => {
  return [formName.charAt(0).toLowerCase() + formName.slice(1), 'Fields'].join('');
};

/**
 * @returns true if string contains special characters except for "." , "_" and 0-9
 * and not a member of the resveredName set
 */
export const isValidVariableName = (input: string, reservedNames?: Set<string>): boolean => {
  const preCheck = reservedNames ? !reservedNames.has(input) : true;
  return /^[a-zA-Z_$][a-zA-Z_$0-9]*$/g.test(input) && preCheck;
};

export const formatFieldName = (input: string): string => camelCase(input.split('.')[0].replace(/[^a-zA-Z_$]/g, '-'));

function elementIsInput(element: FormDefinitionElement): element is FormDefinitionInputElement {
  return element.componentType !== 'Text' && element.componentType !== 'Divider' && element.componentType !== 'Heading';
}

// create mapping for field name collisions
export function generateUniqueFieldName(name: string, sanitizedFieldNames: Set<string>) {
  let sanitizedFieldName = isValidVariableName(name, sanitizedFieldNames) ? name : formatFieldName(name);
  let count = 1;
  if (sanitizedFieldNames.has(sanitizedFieldName.toLowerCase()) && !name.includes('.')) {
    let prospectiveNewName = sanitizedFieldName + count;
    while (sanitizedFieldNames.has(prospectiveNewName.toLowerCase())) {
      count += 1;
      prospectiveNewName = sanitizedFieldName + count;
    }
    sanitizedFieldName = prospectiveNewName;
  }
  sanitizedFieldNames.add(sanitizedFieldName.toLowerCase());
  return sanitizedFieldName !== name.split('.')[0] && sanitizedFieldName;
}

export const mapFormMetadata = (form: StudioForm, formDefinition: FormDefinition): FormMetadata => {
  const sanitizedFieldNames = new Set<string>(reservedWords);
  return {
    id: form.id,
    name: form.name,
    formActionType: form.formActionType,
    dataType: form.dataType,
    layoutConfigs: formDefinition.form.layoutStyle,
    labelDecorator: formDefinition.form.labelDecorator,
    fieldConfigs: Object.entries(formDefinition.elements).reduce<Record<string, FieldConfigMetadata>>(
      (configs, [name, element]) => {
        if (!elementIsInput(element)) {
          return configs;
        }

        const updatedConfigs = configs;
        const metadata: FieldConfigMetadata = {
          validationRules: [],
          componentType: element.componentType,
          studioFormComponentType: 'studioFormComponentType' in element ? element.studioFormComponentType : undefined,
          isArray: element.isArray,
          dataType: element.dataType,
          relationship: element.relationship,
        };
        if ('validations' in element && element.validations) {
          metadata.validationRules = element.validations.map<FieldValidationConfiguration>((validation) => {
            const updatedValidation = validation;
            delete updatedValidation.immutable;
            return updatedValidation;
          });
        }

        if ('valueMappings' in element) {
          metadata.valueMappings = element.valueMappings;
        }

        // we add the name of the model as a resvered word
        if (form.dataType.dataSourceType === 'DataStore') {
          sanitizedFieldNames.add(form.dataType.dataTypeName);
        }
        const sanitizedFieldName = generateUniqueFieldName(name, sanitizedFieldNames);
        if (sanitizedFieldName) {
          metadata.sanitizedFieldName = sanitizedFieldName;
        }

        updatedConfigs[name] = metadata;
        return updatedConfigs;
      },
      {},
    ),
  };
};
