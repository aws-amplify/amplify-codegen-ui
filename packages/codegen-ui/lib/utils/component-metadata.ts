/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import {
  StudioComponent,
  StudioComponentChild,
  StudioComponentDataConfiguration,
  StudioComponentEvent,
  StudioComponentProperty,
  StudioComponentPropertyBinding,
  StateReference,
  FormMetadata,
  GenericDataSchema,
} from '../types';
import { StateReferenceMetadata, computeStateReferenceMetadata } from './state-reference-metadata';

export type ComponentMetadata = {
  hasAuthBindings: boolean;
  requiredDataModels: string[];
  stateReferences: StateReferenceMetadata[];
  componentNameToTypeMap: Record<string, string>;
  formMetadata?: FormMetadata;
  dataSchemaMetadata?: GenericDataSchema;
};

/**
 * Component Structure Functions
 */

// Check top-level binding properties, then call helper on the chilren.
export function computeComponentMetadata(component: StudioComponent): ComponentMetadata {
  const { bindingProperties, collectionProperties } = component;
  const componentMetadataWithoutDataDependencies = dedupeComponentMetadata(
    reduceComponentMetadata([
      ...(bindingProperties ? Object.values(bindingProperties).map(computeBindingPropertyMetadata) : []),
      ...(collectionProperties ? Object.values(collectionProperties).map(computeCollectionPropertyMetadata) : []),
      computeComponentMetadataHelper(component),
    ]),
  );

  return {
    ...componentMetadataWithoutDataDependencies,
    stateReferences: computeStateReferenceMetadata(
      component,
      componentMetadataWithoutDataDependencies.stateReferences.map(({ reference }) => reference),
    ),
  };
}

// Check names, properties, events, and recurse through children.
function computeComponentMetadataHelper(component: StudioComponent | StudioComponentChild): ComponentMetadata {
  return reduceComponentMetadata([
    generateNameMappingMetadata(component.name, component.componentType),
    ...(component.properties ? Object.values(component.properties).map(computePropertyMetadata) : []),
    ...(component.events ? Object.values(component.events).map(computeEventMetadata) : []),
    ...(component.children ? component.children.map(computeComponentMetadataHelper) : []),
  ]);
}

/**
 * Utility Functions
 */

// Because object equality won't catch dupes here, we'll map by component Name, dedupe
// properties, then unroll to a list.
// Dont dedupe setters though, since they'll all be their own ref
function dedupeStateReferences(stateReferences: StateReferenceMetadata[]): StateReferenceMetadata[] {
  const stateReferenceMap: Record<string, Set<string>> = {};
  stateReferences
    .filter(({ reference }) => !('set' in reference) || reference.set === null || reference.set === undefined)
    .forEach(({ reference: { componentName, property } }) => {
      if (!(componentName in stateReferenceMap)) {
        stateReferenceMap[componentName] = new Set([]);
      }
      stateReferenceMap[componentName].add(property);
    });
  return Object.entries(stateReferenceMap)
    .flatMap(([componentName, properties]) =>
      [...properties].map((property) => {
        const dataDependencies: string[] = [];
        return { reference: { componentName, property }, dataDependencies };
      }),
    )
    .concat(stateReferences.filter(({ reference }) => 'set' in reference && reference.set));
}

function dedupeComponentMetadata(componentMetadata: ComponentMetadata): ComponentMetadata {
  const { hasAuthBindings, requiredDataModels, stateReferences, componentNameToTypeMap } = componentMetadata;
  return {
    hasAuthBindings,
    requiredDataModels: [...new Set(requiredDataModels)],
    stateReferences: dedupeStateReferences(stateReferences),
    componentNameToTypeMap,
  };
}

function reduceComponentMetadata(componentMetadata: ComponentMetadata[]): ComponentMetadata {
  const mergeMetadata = (lhs: ComponentMetadata, rhs: ComponentMetadata) => {
    return {
      hasAuthBindings: lhs.hasAuthBindings || rhs.hasAuthBindings,
      requiredDataModels: [...lhs.requiredDataModels, ...rhs.requiredDataModels],
      stateReferences: [...lhs.stateReferences, ...rhs.stateReferences],
      componentNameToTypeMap: { ...lhs.componentNameToTypeMap, ...rhs.componentNameToTypeMap },
    };
  };
  return componentMetadata.reduce(mergeMetadata, generateEmptyMetadata());
}

function generateEmptyMetadata(): ComponentMetadata {
  return {
    hasAuthBindings: false,
    requiredDataModels: [],
    stateReferences: [],
    componentNameToTypeMap: {},
  };
}

function generateAuthBindingMetadata(): ComponentMetadata {
  return {
    hasAuthBindings: true,
    requiredDataModels: [],
    stateReferences: [],
    componentNameToTypeMap: {},
  };
}

function generateModelMetadata(model: string): ComponentMetadata {
  return {
    hasAuthBindings: false,
    requiredDataModels: [model],
    stateReferences: [],
    componentNameToTypeMap: {},
  };
}

function generateReferenceMetadata(reference: StateReference): ComponentMetadata {
  return {
    hasAuthBindings: false,
    requiredDataModels: [],
    stateReferences: [{ reference, dataDependencies: [] }],
    componentNameToTypeMap: {},
  };
}

function generateNameMappingMetadata(componentName: string | undefined, componentType: string): ComponentMetadata {
  if (componentName) {
    return {
      hasAuthBindings: false,
      requiredDataModels: [],
      stateReferences: [],
      componentNameToTypeMap: { [componentName]: componentType },
    };
  }
  return generateEmptyMetadata();
}

/**
 * Binding Property Functions
 */

function computeCollectionPropertyMetadata(dataConfiguration: StudioComponentDataConfiguration): ComponentMetadata {
  return generateModelMetadata(dataConfiguration.model);
}

function computeBindingPropertyMetadata(bindingProperty: StudioComponentPropertyBinding): ComponentMetadata {
  switch (bindingProperty.type) {
    case 'Event':
      return generateEmptyMetadata();
    case 'Data':
      return generateModelMetadata(bindingProperty.bindingProperties.model);
    case 'Storage':
      return generateEmptyMetadata();
    default:
      if ('type' in bindingProperty) {
        return generateEmptyMetadata();
      }
      throw new Error(`Binding Property ${JSON.stringify(bindingProperty)} could not be parsed for Metadata.`);
  }
}

/**
 * Property Functions
 */

function computePropertyMetadata(property: StudioComponentProperty): ComponentMetadata {
  const meta =
    typeof property !== 'object'
      ? []
      : ([] as ComponentMetadata[])
          .concat('concat' in property && property.concat ? property.concat.map(computePropertyMetadata) : [])
          .concat('userAttribute' in property && property.userAttribute ? [generateAuthBindingMetadata()] : [])
          .concat(
            'condition' in property && property.condition && 'then' in property.condition && property.condition.then
              ? [computePropertyMetadata(property.condition.then)]
              : [],
          )
          .concat(
            'condition' in property && property.condition && 'else' in property.condition && property.condition.else
              ? [computePropertyMetadata(property.condition.else)]
              : [],
          )
          .concat(
            'componentName' in property && property.componentName && 'property' in property && property.property
              ? [generateReferenceMetadata(property as StateReference)]
              : [],
          );

  return reduceComponentMetadata(meta);
}

/**
 * Event Functions
 */

function computeEventMetadata(event: StudioComponentEvent): ComponentMetadata {
  if (!('action' in event) || event.action === undefined || event.action === null) {
    return generateEmptyMetadata();
  }

  switch (event.action) {
    case 'Amplify.Navigation':
      return reduceComponentMetadata(Object.values(event.parameters).map(computePropertyMetadata));
    case 'Amplify.AuthSignOut':
      return reduceComponentMetadata(Object.values(event.parameters).map(computePropertyMetadata));
    case 'Amplify.DataStoreCreateItemAction':
      return reduceComponentMetadata([
        generateModelMetadata(event.parameters.model),
        ...Object.values(event.parameters.fields).map(computePropertyMetadata),
      ]);
    case 'Amplify.DataStoreUpdateItemAction':
      return reduceComponentMetadata([
        generateModelMetadata(event.parameters.model),
        computePropertyMetadata(event.parameters.id),
        ...Object.values(event.parameters.fields).map(computePropertyMetadata),
      ]);
    case 'Amplify.DataStoreDeleteItemAction':
      return reduceComponentMetadata([
        generateModelMetadata(event.parameters.model),
        computePropertyMetadata(event.parameters.id),
      ]);
    case 'Amplify.Mutation':
      return reduceComponentMetadata([
        generateReferenceMetadata(event.parameters.state),
        computePropertyMetadata(event.parameters.state.set),
      ]);
    default:
      throw new Error(`Event ${JSON.stringify(event)} could not be parsed for Metadata.`);
  }
}
