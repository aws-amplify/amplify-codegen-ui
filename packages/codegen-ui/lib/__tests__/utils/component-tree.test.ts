/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { getComponentFromComponentTree } from '../../utils/component-tree';

describe('getComponentFromComponentTree', () => {
  const grandChildComponent = {
    componentType: 'TextField',
    name: 'GrandChildComponent',
    properties: {},
  };
  const childComponent = {
    componentType: 'Flex',
    name: 'ChildComponent',
    properties: {},
    children: [grandChildComponent],
  };
  const component = {
    componentType: 'Flex',
    name: 'Component',
    properties: {},
    bindingProperties: {},
    children: [childComponent],
  };
  test('same as root component', () => {
    expect(getComponentFromComponentTree(component, 'Component')).toEqual(component);
  });

  test('child component', () => {
    expect(getComponentFromComponentTree(component, 'ChildComponent')).toEqual(childComponent);
  });

  test('grandchild component', () => {
    expect(getComponentFromComponentTree(component, 'GrandChildComponent')).toEqual(grandChildComponent);
  });

  test('not found', () => {
    expect(() => getComponentFromComponentTree(component, 'NotFoundComponent')).toThrowErrorMatchingSnapshot();
  });
});
