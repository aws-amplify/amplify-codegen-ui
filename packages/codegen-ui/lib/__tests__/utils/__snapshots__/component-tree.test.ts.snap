// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentFromComponentTree not found 1`] = `"Component NotFoundComponent not found in component tree Component"`;
